feat: Template executor plugins. Fixes #5201 (#7256)
argoproj/argo-workflows: docs/plugins.md

# Plugins

![Plugins](assets/plugins.png)

Plugins allow you to extend Argo Workflows to add new capabilities.

* You don't need to learn Golang, you can write in any language, including Python.
* Simple: a plugin just responds to RPC HTTP requests.
* You can iterate quickly by changing the plugin at runtime.
* You can get your plugin running today, no need to wait 3-5 months for an Argo software release.

[Executor plugins](executor_plugins.md) written and installed by both users and admins.

## Configuration

> Plugins are disabled by default. To enable them, start the controller with `ARGO_PLUGINS=true`, e.g.

Reviewer: Just thinking it out: an allowlist would solve for the situation where an untrusted user may _add_ a plugin ConfigMap, but cannot _modify_ an existing plugin ConfigMap. 

If the malicious user could modify an existing plugin ConfigMap, they'd just add malicious code to the existing plugin rather than creating a new malicious plugin.

> Plugins are disabled by default. To enable them, start the controller with `ARGO_PLUGINS=true`, e.g.

Author: I'm not sure an allow/deny list is needed. Users can add and remove plugins by creating and deleting config maps.

What _might_ be useful is some way to only enable plugins for some workflows, e.g. by label selector.

```yaml
apiVersion: apps/v1
kind: Deployment
metadata:
  name: workflow-controller
spec:
  template:
    spec:
      containers:
        - name: workflow-controller
          env:
            - name: ARGO_PLUGINS
              value: "true"
```

## Considerations

### Security

> No security context is enforced, but is strong encouraged you configure your container not to run as root.

Reviewer: ```suggestion
No security context is enforced, but it is strongly that encouraged you configure your container not to run as root.
```

Plugins can run at an non-localhost address, but this is very strongly discouraged. Within a pod, network traffic cannot

> be observed.

Reviewer: Ditto above. If we're going to support this mode, I think it should be first-class.

### Resources

> No resource requests or limits are enforced (as they move override system defaults), but you are strongly encouraged to

Author: "may"

set these.

### Failure Modes

A plugin may fail as follows:

* Connection/socket problems.

> * Timeout (1s for controller plugins, 30s for executor plugins).

Reviewer: ```suggestion
* Timeout (30s for executor plugins).
```

No controller plugins yet.

* Transient error.
* 4xx or 5xx error:
    * 404 error - endpoint will not be invoked again.
    * 503 error - considered a transient error.

> * Multiple invocations of the same plugin take too long.

Reviewer: I'm not sure what this means.

> Transient errors are retried, all other errors are considered fatal.

Reviewer: Based on the list above, I'm not 100% sure what transient errors are. I know that 503s are transient, but connection/socket, timeout, and multiple invocations... are those transient?

Author: Any error that might be transient, i.e. trying again shortly might succeed. This include network congestion.

Reviewer 2: Besides the code, do we have a formal definition of transient errors documented somewhere? We probably should

> Transient errors are retried, all other errors are considered fatal.

Author: no plan for v1

Fatal errors are typically contained as follows:

> * For node lifecycle hooks, the node will error. The workflow therefore may fail.

Reviewer: Does this PR include node lifecycle hooks? I think this is the first mention of them in this doc.

> * Other errors will result in an errored workflow.

Reviewer: ```suggestion
* Other errors will result in a failed workflow.
```

I think that terminology's probably more consistent with other docs.

Author: done

> ### Performance Is Important

Reviewer 3: Do we like to give the user some suggestion on how to make it performance rather than just stating that is matters?

Reviewer 4: I might be wrong but I feel like that's out of the scope of this document since plugins are kind of arbitrary.

> Consider a workflows with 100k nodes, and then consider you have 5 plugins:

Reviewer: ```suggestion
Consider a workflow with 100k nodes, and then consider you have 5 plugins:
```

Author: done
